Modules/_pickle.c

```diff
@@ -1292,13 +1358,28 @@ _Unpickler_ReadFromFile(UnpicklerObject *self, Py_ssize_t n)
                     return n;
             }
         }
-        len = PyLong_FromSsize_t(n);
+        Py_ssize_t cursize = Py_MIN(n, SAFE_BUF_SIZE);
+        len = PyLong_FromSsize_t(cursize);
         if (len == NULL)
             return -1;
         data = _Pickle_FastCall(self->read, len);
+        if (data == NULL) {
+            return -1;
+        }
+        while (cursize < n) {
+            Py_ssize_t prevsize = cursize;
+            cursize += Py_MIN(cursize, n - cursize);
+            if (_PyBytes_Resize(&data, cursize) < 0) {
```

Comment:
Doing this in a loop is O(N^2) just like in the pure Python code. We need to use the list of chunks & join approach to avoid that.